### PR TITLE
etcd using asset builder

### PR DESCRIPTION
**Pull request by chrislovecnm**
Updating etcd to use asset builder

### CODE REVIEW

File: pkg/model/components/etcd.go [hunk 1]
```diff
@@ -32,13 +34,31 @@ var _ loader.OptionsBuilder = &EtcdOptionsBuilder{}
...
 	// @check the version are set and if not preset the defaults
 	for _, x := range spec.EtcdClusters {
 		// @TODO if nothing is set, set the defaults. At a late date once we have a way of detecting a 'new' cluster
 		// we can default all clusters to v3
 		if x.Version == "" {
 			x.Version = "2.2.1"
 		}
+
+		if imageVersion != "" {
+			if imageVersion != x.Version {
```
Comment (chrislovecnm) [inline pkg/model/components/etcd.go]: Here is the odd thing @gambol99 - I think we could technically end up with multiple different versions of etcd here.  Please tell me I am wrong.  I am wondering if we should move the version up a level in the structs.

**Comment by chrislovecnm** [comment]
/assign @KashifSaadat 

I am not loving that nodeup has to call assets builder, but maybe we refactor later?

File: nodeup/pkg/model/protokube.go [hunk 2]
```diff
@@ -219,11 +220,23 @@ func (t *ProtokubeBuilder) ProtokubeFlags(k8sVersion semver.Version) *ProtokubeF
 	f := &ProtokubeFlags{
 		Channels:      t.NodeupConfig.Channels,
 		Containerized: fi.Bool(true),
-		EtcdImage:     s(fmt.Sprintf("gcr.io/google_containers/etcd:%s", imageVersion)),
 		LogLevel:      fi.Int32(4),
 		Master:        b(t.IsMaster),
 	}
 
+	// TODO this is dupicate code with etcd model
+	image := fmt.Sprintf("gcr.io/google_containers/etcd:%s", imageVersion)
```
Comment (KashifSaadat) [inline nodeup/pkg/model/protokube.go]: Could we make this configurable to allow users to specify either the full address or version within their Cluster Spec? Similar to how it works for `KubernetesVersion`.

**Comment by KashifSaadat** [comment]
One minor comment, otherwise LGTM and is functional 👍

File: pkg/model/components/etcd.go (hunk 1 above)
Comment (gambol99) [reply pkg/model/components/etcd.go]: logic seems right ... on first iteration set and every other has to equal .. unless i'm missing something obvious  :-)

Comment (gambol99) [reply pkg/model/components/etcd.go]: initial the version was set in the outer struct, i recall @justinsb want it to be per etcd cluster though ..

Comment (gambol99) [reply pkg/model/components/etcd.go]: the multiple version was protected on the spec validation https://github.com/kubernetes/kops/blob/master/pkg/apis/kops/validation/legacy.go#L473-L495

File: nodeup/pkg/model/protokube.go (hunk 2 above)
Comment (chrislovecnm) [reply nodeup/pkg/model/protokube.go]: That is exactly what https://github.com/kubernetes/kops/blob/master/pkg/apis/kops/cluster.go#L176 does. The assests builder uses the container registry setting.

kops will upload the dependencies to you registry during the assests phase and then your cluster can use them.

This work is me wiring everything together.

Comment (KashifSaadat) [reply nodeup/pkg/model/protokube.go]: Ah I missed that, ok perfect! 👍

File: pkg/model/components/etcd.go (hunk 1 above)
Comment (chrislovecnm) [reply pkg/model/components/etcd.go]: If I am reading that code it does not validate it.

File: nodeup/pkg/model/protokube.go [hunk 3]
```diff
@@ -219,11 +220,23 @@ func (t *ProtokubeBuilder) ProtokubeFlags(k8sVersion semver.Version) *ProtokubeF
...
+	assets := assets.NewAssetBuilder(nil)
+	remapped, err := assets.RemapImage(image)
+	if err != nil {
+		glog.Errorf("unable to remap container %q: %v", image, err)
```
Comment (justinsb) [inline nodeup/pkg/model/protokube.go]: What is the reason to fall back?  Just return `fmt.Errorf` I think

File: pkg/model/components/etcd.go (hunk 1 above)
Comment (justinsb) [reply pkg/model/components/etcd.go]: Can we move this to validation?

File: pkg/model/components/etcd.go [hunk 4]
```diff
@@ -32,13 +34,31 @@ var _ loader.OptionsBuilder = &EtcdOptionsBuilder{}
...
+			}
+		} else {
+			imageVersion = x.Version
+		}
+	}
+
+	// TODO we do not have a API place to put this.  Do we use version or get another
+	// TODO API value?
+	// TODO code is duplicated in nodeup
+	image := fmt.Sprintf("gcr.io/google_containers/etcd:%s", imageVersion)
```
Comment (justinsb) [inline pkg/model/components/etcd.go]: This is validating that the assetbuilder is working?  Maybe we should explicitly put the etcd image into the nodeup spec.

Comment (chrislovecnm) [reply pkg/model/components/etcd.go]: no this is building the URL for image builder ... 

Yes, we should put it in the nodeup spec.  This PR or another?  I would say in another PR.  But shouldn't it be one container in the etcd spec??  I am seriously confused why we have the version where we have it :(

File: nodeup/pkg/model/protokube.go (hunk 3 above)
Comment (chrislovecnm) [reply nodeup/pkg/model/protokube.go]: did not want to refactor method.  Keep the changes small.  I can refactor to return err

File: pkg/model/components/etcd.go (hunk 1 above)
Comment (chrislovecnm) [reply pkg/model/components/etcd.go]: @gambol99 I will back out the code, can you guys move it to validation?  But, I am really confused why we have the version where it is.  Oh well :)  I can do that validation later, but I have a long long list of things on my plate.

File: pkg/model/components/etcd.go (hunk 4 above)
Comment (chrislovecnm) [reply pkg/model/components/etcd.go]: Do we want the etcd container in the clusterspec and nodeup api?  Or can we do that in another PR?

**Comment by chrislovecnm** [comment]
@justinsb open question is do you want me to fix the etcd API stuff in this PR or another.  I would recommend that we have the etcd container URL in the cluster api.  In this struct https://github.com/kubernetes/kops/blob/master/pkg/apis/kops/cluster.go#L286.  That would also be set in the nodeup API as well.

File: nodeup/pkg/model/protokube.go (hunk 3 above)
Comment (chrislovecnm) [reply nodeup/pkg/model/protokube.go]: refactored

File: pkg/model/components/etcd.go (hunk 4 above)
Comment (chrislovecnm) [reply pkg/model/components/etcd.go]: We have the asset builder here do you want to use that or an API value?

File: nodeup/pkg/model/protokube.go [hunk 5]
```diff
@@ -211,19 +215,30 @@ type ProtokubeFlags struct {
...
 		Channels:      t.NodeupConfig.Channels,
 		Containerized: fi.Bool(true),
-		EtcdImage:     s(fmt.Sprintf("gcr.io/google_containers/etcd:%s", imageVersion)),
 		LogLevel:      fi.Int32(4),
 		Master:        b(t.IsMaster),
 	}
 
+	// TODO this is dupicate code with etcd model
+	image := fmt.Sprintf("gcr.io/google_containers/etcd:%s", imageVersion)
+	assets := assets.NewAssetBuilder(nil)
```
Comment (justinsb) [inline nodeup/pkg/model/protokube.go]: Don't we need to pass in the Assets from the cluster spec (and not `nil`)?

**Comment by chrislovecnm** [comment]
@justinsb PTAL

**Pull request merged**